### PR TITLE
VDS-3182: Add validation for physical product import

### PR DESCRIPTION
**Pull request by asvishnyakov**
Description: 

--

QA-test: VDS-2916

Demo-test: VDS-2917

Download artifact URL: https://vc3prerelease.blob.core.windows.net/packages/VirtoCommerce.CatalogExportImport_1.1.0-pr-26-6a704d3f.zip

### CODE REVIEW

File: src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs [hunk 1]
```diff
@@ -1,17 +1,134 @@
...
+using VirtoCommerce.CatalogExportImportModule.Core;
 using VirtoCommerce.CatalogExportImportModule.Core.Models;
+using static VirtoCommerce.CatalogExportImportModule.Data.Helpers.ValidationExtensions;
 
 namespace VirtoCommerce.CatalogExportImportModule.Data.Validation
 {
-    public class ImportPhysicalProductValidator : AbstractValidator<ImportRecord<CsvPhysicalProduct>[]>
+    public class ImportPhysicalProductValidator: AbstractValidator<ImportRecord<CsvPhysicalProduct>>
     {
+        private static readonly char[] ProductSkuIllegalCharacters = { '$', '+', ';', '=', '%', '{', '}', '[', ']', '|', '@', '~', '!', '^', '*', '&', '(', ')', '<', '>' };
```
Comment (trueboroda) [inline src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs:11]: it should be the const on a platform level.

File: src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs [hunk 2]
```diff
@@ -1,17 +1,134 @@
...
+            RuleFor(record => record.Record.Gtin)
+                .MaximumLength(64)
+                .WithExceededMaxLengthCodeAndMessage("Gtin", 64)
+                .WithImportState();
+
+            RuleFor(record => record.Record.PackageType)
+                .MaximumLength(254)
+                .WithExceededMaxLengthCodeAndMessage("Package Type", 254)
+                .WithImportState()
+                .Must((record, packageType, context) =>
```
Comment (trueboroda) [inline src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs:66]: Use a not cascade mode for this rule.

File: src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs [hunk 3]
```diff
@@ -1,17 +1,134 @@
...
+
+            RuleFor(record => record.Record.ManufacturerPartNumber)
+                .MaximumLength(128)
+                .WithExceededMaxLengthCodeAndMessage("Manufacturer Part Number", 128)
+                .WithImportState();
+
+            RuleFor(record => record.Record.MeasureUnit)
+                .MaximumLength(32)
+                .WithExceededMaxLengthCodeAndMessage("Measure Unit", 32)
+                .WithImportState()
```
Comment (trueboroda) [inline src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs:84]: Use not cascade mode for this rule

**Review by trueboroda — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs [hunk 4]
```diff
@@ -1,17 +1,134 @@
...
+                .Must((record, measureUnit, context) =>
+                {
+                    var availableMeasureUnits = (string[])context.ParentContext.RootContextData[ModuleConstants.ValidationContextData.AvailableMeasureUnits];
+                    return availableMeasureUnits.Contains(measureUnit);
+                })
+                .When(record => !string.IsNullOrEmpty(record.Record.MeasureUnit))
+                .WithInvalidValueCodeAndMessage("Measure Unit")
+                .WithImportState();
+
+            RuleFor(record => record.Record.WeightUnit)
```
Comment (trueboroda) [inline src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs:94]: Use not cascade mode for this rule

File: src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs [hunk 5]
```diff
@@ -1,17 +1,134 @@
...
+                .Must((record, weightUnit, context) =>
+                {
+                    var availableWeightUnits = (string[])context.ParentContext.RootContextData[ModuleConstants.ValidationContextData.AvailableWeightUnits];
+                    return availableWeightUnits.Contains(weightUnit);
+                })
+                .When(record => !string.IsNullOrEmpty(record.Record.WeightUnit))
+                .WithInvalidValueCodeAndMessage("Weight Unit")
+                .WithImportState();
+
+            RuleFor(record => record.Record.TaxType)
```
Comment (trueboroda) [inline src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs:108]: Use not cascade mode for this rule

**Review by trueboroda — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs (hunk 1 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs:11]: https://virtocommerce.atlassian.net/browse/PT-5208 that's all what I can do at the moment :)

File: src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs (hunk 2 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs:66]: Fixed

File: src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs (hunk 3 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs:84]: Fixed

File: src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs (hunk 4 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs:94]: Fixed

File: src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs (hunk 5 above)
Comment (asvishnyakov) [reply src/VirtoCommerce.CatalogExportImportModule.Data/Validation/ImportPhysicalProductValidator.cs:108]: Fixed

**Review by trueboroda — APPROVED** [review]

**Pull request merged**